### ISSUE TITLE
addCleanup method on TestCase to allow for easier tear down.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2610]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#2610|
|Type|enhancement|
|Created|2007-04-29 05:39:21Z|

bzrlib has a neat feature on their TestCase subclass called `addCleanup`. We should emulate this feature.

<details><summary>Searchable metadata</summary>

```
trac-id__2610 2610
type__enhancement enhancement
reporter__jml jml
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1177825161000000 1177825161000000
changetime__1179371682000000 1179371682000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml removed owner|
|-|-|

Ready for review in `addCleanup-2610`

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

Could you add some examples in the twisted test suite where it's actually useful? I think I understand what it does but I don't see what it brings against tearDown. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_3"></a>|@jml commented|
|-|-|

Off the top of my head:

 1. `addCleanup` keeps the teardown code physically close to the setup code.
 1. `addCleanup` makes it easy to have factory methods in your test classes for things which require substantial dismantling.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_4"></a>|@jml commented|
|-|-|

See r20104

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @jml|
|-|-|

Please follow the coding standard.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_6"></a>|@jml commented|
|-|-|

Normally, I find your code reviews extremely valuable and helpful. This one is an exception.

It would be more helpful if you cited specific coding standard violations, or at least which kinds of coding standard violations are prevalent throughout this branch.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

 * All the methods you've added should include docstrings in the format : 
```
"""
doc
"""
```
 * Test methods names should only have the first underscore. For example you may change test_addCleanupCalled_in_reverse_order to test_addCleanupReverseOrder
 * Unused import in test_tests
 * Please add copyright statement in test_tests, and update the one in unittest.
 * The code in _runCleanups is hard to read, and its intend is not very clear. I don't think fireOnOneErrback and consumeErrors are compatible options (the first overriding the second). We probably want to run all the cleanup functions, regardless if another fail, so consumeError seems appropriate. But I feel that it would be a good place for a deferGenerator.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_9"></a>|@jml set owner to @therve|
|-|-|

Fixed up the copyright and coding standard issues. 

I've moved the 'sequentially run functions which may or may not return deferreds' logic to a function in `t.internet.defer` called `runSequentially` and modified it to use `deferredGenerator`. It is clearer.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

Humm... I really like runSequentially, and the code in _runCleanups is indeed much clearer. But I feel that this change might be big enought to deserve its own branch. What do you think?

As I understand that you may not want to postpone this forever, you can remove runSequentially from here, add a TODO in _runCleanups, then merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to @jml|
|-|-|

Last comment: deferRunCleanups and _cbDeferRunCleanups are missing docstrings.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_12"></a>|@jml set status to `closed`|
|-|-|

(In [20290]) Add `TestCase.addCleanup` and `defer.runSequentially`.

 * Author: jml
 * Reviewer: therve
 * Fixes [#2610](https://github.com/twisted/twisted/issues/2610)

`addCleanup` is an easy way to add cleanup routines to unit tests. It's
particularly useful for factory methods which return objects that require some
sort of cleanup (such as an explicit disconnect). Because this is Twisted,
callables passed to `addCleanup` can return Deferreds.

`runSequentially` takes a sequence of nullary callables and runs them
sequentially. If a callable returns a `Deferred`, `runSequentially` will not
run the next callable until the callback chain of the `Deferred` has been
completed.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
